pagerduty/resource_pagerduty_slack_connection.go

```diff
@@ -0,0 +1,309 @@
+		} else if slackConn != nil {
+			d.SetId(slackConn.ID)
+			d.Set("workspace_id", slackConn.WorkspaceID)
+		}
+		return nil
+	})
+	if retryErr != nil {
+		time.Sleep(2 * time.Second)
+		return retryErr
+	}
+	return resourcePagerDutySlackConnectionRead(d, meta)
+}
+
+func resourcePagerDutySlackConnectionRead(d *schema.ResourceData, meta interface{}) error {
+	config := &pagerduty.Config{
+		Token:   os.Getenv("PAGERDUTY_USER_TOKEN"),
+		BaseURL: "https://app.pagerduty.com",
+	}
+	client, err := pagerduty.NewClient(config)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Reading PagerDuty slack connection %s", d.Id())
+
+	workspaceID := d.Get("workspace_id").(string)
+	log.Printf("[DEBUG] Read Slack Connection: workspace_id %s", workspaceID)
+
+	retryErr := resource.Retry(2*time.Minute, func() *resource.RetryError {
+		if slackConn, _, err := client.SlackConnections.Get(workspaceID, d.Id()); err != nil {
+			return resource.RetryableError(err)
+		} else if slackConn != nil {
+			d.Set("source_id", slackConn.SourceID)
+			d.Set("source_name", slackConn.SourceName)
+			d.Set("source_type", slackConn.SourceType)
+			d.Set("channel_id", slackConn.ChannelID)
+			d.Set("channel_name", slackConn.ChannelName)
+			d.Set("notification_type", slackConn.NotificationType)
+			d.Set("config", flattenConnectionConfig(slackConn.Config))
+		}
+		return nil
+	})
+
+	if retryErr != nil {
+		time.Sleep(2 * time.Second)
+		return retryErr
+	}
+
+	return nil
+}
+
+func resourcePagerDutySlackConnectionUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := &pagerduty.Config{
+		Token:   os.Getenv("PAGERDUTY_USER_TOKEN"),
+		BaseURL: "https://app.pagerduty.com",
+	}
+	client, err := pagerduty.NewClient(config)
+	if err != nil {
+		return err
+	}
+
+	slackConn, err := buildSlackConnectionStruct(d)
+	if err != nil {
+		return err
+	}
+	log.Printf("[INFO] Updating PagerDuty slack connection %s", d.Id())
+
+	if _, _, err := client.SlackConnections.Update(slackConn.WorkspaceID, d.Id(), slackConn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourcePagerDutySlackConnectionDelete(d *schema.ResourceData, meta interface{}) error {
+	config := &pagerduty.Config{
+		Token:   os.Getenv("PAGERDUTY_USER_TOKEN"),
+		BaseURL: "https://app.pagerduty.com",
+	}
+	client, err := pagerduty.NewClient(config)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Deleting PagerDuty slack connection %s", d.Id())
+	workspaceID := d.Get("workspace_id").(string)
+
+	if _, err := client.SlackConnections.Delete(workspaceID, d.Id()); err != nil {
+		return err
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func expandConnectionConfig(v interface{}) pagerduty.ConnectionConfig {
+	c := v.([]interface{})[0].(map[string]interface{})
+
+	var config pagerduty.ConnectionConfig
+
+	config = pagerduty.ConnectionConfig{
+		Events:     expandConfigList(c["events"].([]interface{})),
+		Priorities: expandConfigList(c["priorities"].([]interface{})),
+		Urgency:    nil,
```

Comment:
Does this need to be nil? 🙂

Reply:
Yeah, the API won't allow it to be a blank string. It's either a valid value or null.